scripts/tests/test_safe_template_linter.py

```diff
@@ -335,3 +434,137 @@ def _validate_expression(self, template_string, expression, expected_expression)
         end_index = expression['end_index']
         self.assertEqual(template_string[start_index:end_index + 1], expected_expression)
         self.assertEqual(expression['expression'], expected_expression)
+
+
+@ddt
+class TestUnderscoreTemplateLinter(TestCase):
+    """
+    Test UnderscoreTemplateLinter
+    """
+
+    def test_check_underscore_file_is_safe(self):
+        """
+        Test _check_underscore_file_is_safe with safe template
+        """
+        linter = UnderscoreTemplateLinter()
+        results = FileResults('')
+
+        template = textwrap.dedent("""
+            <%- gettext('Single Line') %>
+
+            <%-
+                gettext('Multiple Lines')
+            %>
+        """)
+
+        linter._check_underscore_file_is_safe(template, results)
+
+        self.assertEqual(len(results.violations), 0)
+
+    def test_check_underscore_file_is_not_safe(self):
+        """
+        Test _check_underscore_file_is_safe with unsafe template
+        """
+        linter = UnderscoreTemplateLinter()
+        results = FileResults('')
+
+        template = textwrap.dedent("""
+            <%= gettext('Single Line') %>
+
+            <%=
+                gettext('Multiple Lines')
+            %>
+        """)
+
+        linter._check_underscore_file_is_safe(template, results)
+
+        self.assertEqual(len(results.violations), 2)
+        self.assertEqual(results.violations[0].rule, Rules.underscore_not_escaped)
+        self.assertEqual(results.violations[1].rule, Rules.underscore_not_escaped)
+
+    @data(
+        {
+            'template':
+                '<% // safe-lint: disable=underscore-not-escaped %>\n'
+                '<%= message %>',
+            'is_disabled': True,
+        },
+        {
+            'template':
+                '<% // safe-lint: disable=another-rule,underscore-not-escaped %>\n'
+                '<%= message %>',
+            'is_disabled': True,
+        },
+        {
+            'template':
+                '<% // safe-lint: disable=another-rule %>\n'
+                '<%= message %>',
+            'is_disabled': False,
+        },
+        {
+            'template':
+                '\n<% // safe-lint: disable=underscore-not-escaped %>\n'
+                '<%= message %>\n'
+                '<%- message %>',
+            'is_disabled': True,
+        },
+        {
+            'template':
+                '<%= message %><% // safe-lint: disable=underscore-not-escaped %>\n'
+                '<%- message %>',
+            'is_disabled': True,
+        },
+        {
+            'template':
+                '<%= message %>\n'
+                '<% // safe-lint: disable=underscore-not-escaped %>',
+            'is_disabled': False,
+        },
+    )
+    def test_check_underscore_file_disable_rule(self, data):
```

Comment:
If the heuristic @nedbat mentioned above gets implemented, it ought to be tested as well